README
Introduction:
=============

Loudmouth is a lightweight and easy-to-use C library for programming with the
Jabber protocol. It's designed to be easy to get started with and yet
extensible to let you do anything the Jabber protocol allows.

The goal is to fully support Linux, *BSD, Mac OSX and Windows. It has
currently only been tested on Linux (Red Hat 8.0, Debian Unstable) and FreeBSD
that I know of. If someone with access to other machines can provide
feedback I would be most grateful.


Requirements:
=============

Glib >= 2.4.0:
ftp://ftp.gtk.org/pub/gtk/v2.2/

gtk-doc (optional, if you want documentation built):
ftp://ftp.gnome.org/pub/GNOME/sources/gtk-doc/1.0

If you want SSL-support you'll also need:
GnuTLS >= 1.2.0 with libtasn support.

Building:
=========

./configure --prefix=
make
make install


Debugging:
==========

You can get debug information printed to STDOUT by using the environment 
variable LM_DEBUG.

export LM_DEBUG=<keyword>

Where keyword is one or more (separated by :) of the following:

* "NET": Shows all the data sent over the network.
* "PARSER": Shows debug output from the parser.
* "VERBOSE": Shows various verbose output.
* "ALL": Enables all output.


Questions, remarks, bug reports:
================================

See the web page for more information and issue reporting.

http://projects.imendio.com/loudmouth

Questions and others can be sent to me at:
email: micke@imendio.com
jabber: micke@imendio.com


NEWS
** Unused, please look at the release notes **




Changes in 1.2.0:
---------------
* This is the same as 1.1.5 but marked as stable and released as 1.2.0.
* For a complete list of changes since 1.0.5 look at the changes in 1.1.1-1.1.5

Changes in 1.1.5:
-----------------
* Fixed leak when closing a connection, LM-59
* Escape the resource string (fixes problems with iChat contacts), LM-48
* Stale incoming source destroyed/unreffed, LM-60
* Fixed a leak in the synchronous layer, LM-66
* Fixed a couple of warning when building on 64 bit alpha, LM-41.

Changes in 1.1.4:
-----------------
* Generate better message IDs, LM-56
* Fixed a leaking proxy object, LM-51
* Fixed random bugs on connection errors, LM-58
* Removed a warning when using proxy, LM-50.

Changes in 1.1.3:
-----------------
* Infinite reading of 0 bytes when disconnected using gnutls, LM-45
* Loudmouth exports some internal SHA* functions, LM-47
* Fix auto* warning, LM-46

Changes in 1.1.2:
-----------------
* Support for internationalised domain names, LM-42
* Fixed issue where no disconnect signal was emitted when connection closed 
  during opening, LM-43
* Plugged a memory leak in LmConnection, LM-39
* Fixed warnings due to uninitialized value, LM-35
* Fixed a reentrancy-related issue with LmConnection, LM-37

Changes in 1.0.3:
-----------------
* Quick build fix release when building without SSL, LM-25.

Changes in 1.0.2:
-----------------
* Fixed issue when opening in blocking mode failed, LM-15
* Unregister sources when connection fails to avoid later segfault, LM-16
* Fixed an issue where large messages would get trunkated, LM-17

Changes in 1.0.1:
-----------------
* Unref contexts when freeing the connection          (Mikael Hallendal)

Changes in 1.0:
---------------
* Fix crash when connection goes down using SSL, LM-3 (Mikael Hallendal)
* Fixed some GCC 4.0 warnings                         (Mikael Hallendal)
* Some API fixups for consistency in LmSLL            (Mikael Hallendal)
* Don't try to setup SSL if it's not supported        (Martyn Russel)
* Documentation fixes                                 (Mikael Hallendal)
* State fixes when using blocking API, LM-7           (Mikael Hallendal)
* Check that connection is open before using it, LM-6 (Mikael Hallendal)
* Fix infinite loop in blocking open if failed, LM-8  (Mikael Hallendal)
 
Changes in 0.90:
----------------
* Two crashers for ssl read fixed (Micke, Daniel Lavalliere)
* Fixed a couple of leaks when freeing the connection (Daniel Lavalliere)
* A couple of thread related issues solved (Micke, Daniel Lavalliere)
* Work with latest GnuTLS (foser)
* Improved the example tools some (Micke)

Changes in 0.17.2:
------------------
* Unset proxy by setting it to NULL.
* Fixes for 64 bit architectures.
* Fixed an issue when not using default main context.
* Fixed UTF-8 issue that lead to a few weird problems in Gossip.

Changes in 0.17.1:
------------------
* Fixed problem when not setting a keep alive rate.

Changes in 0.17:
----------------
* Compile fixes (Sjoerd Simons
* Depends on GLib 2.4
* Fixed memory leak
* Added support to work tunnel, using another JID than the connect server.
* Support for keep alive packages

Changes in 0.16:
----------------
* Support for giving the context to which an LmConnection should run in
  (should solve issues when wanting to run several LmConnections in various
   threads)
* HTTP proxy support
* Memory leak fix
* Distribute the win32-config.h file needed to build on Windows.
 
Changes in 0.15.1:
------------------
* Require GnuTLS >= 1.0.0

Changes in 0.15:
----------------
* Better Win32 instructions.
* Fixed bug when disconnecting and then connecting again using the same 
  LmConnection.
* Documentation typos and errors fixed by Ross Burton.
* Better SSL support, now gives the application the possibility to decide
  if it wants to connect on bogus SSL certificates.
* The SSL API changed a bit, take a look at the API docs.
* Fixed timestamp error on FreeBSD.

Changes in 0.14.1:
------------------
* Patch from foser@gentoo.org for a problem when using GnuTLS.

Changes in 0.14:
----------------
* Added information about building on Windows
* Bug when using GnuTLS fixed.
* Fixed a bug that could cause replies to be overlooked when using 
  blocking mode.
* Added IPv6-support, patch from Sjoerd Simons.
* Made it easier to build RPMs without SSL-support.
* Fixed a memleak when destroying a connection.

Changes in 0.13.2:
------------------
* Don't crash when we receive </stream:stream>

Changes in 0.13.1:
------------------
* Don't require GnuTLS to build

Changes in 0.13:
----------------
* Fixed bug in the sub type checking code for messages  
* Fixed bug that hang the app when disconnected uncleanly
* Fixed autoconf problems with gnutls                   
* Disconnect cleanly                                   
* Fix sha implementation to work on Linux/PPC         

Changes in 0.12:
----------------
* Added support for cancelling messages waiting for reply
* Now checks for supported auth methods.
* Fixed RPM's to require GnuTLS.

Changes in 0.11:
----------------
* SSL-support, patch by Bartosz Zapalowski

Changes in 0.10.1:
------------------
* Build fix for *BSD. Reported by Ralph Meijer.

Changes in 0.10:
----------------
* Fixed case bug in message nodes. Making it not treat the XML as case 
  dependant.
* Added lm_message_node_[get|set]_raw_mode that determines if
  the node value should be escaped or not.
* Return correct type from lm_message_get_sub_type.
* Win32 build fixes. Sent in by Vince Negri.

Changes in 0.9:
---------------
* Build fixes Solaris
* Non async operations
* API additions to ease binding efforts
* Added Debug output support
* Bug fixes
* API changes (see below).
* Support for disconnect and disconnect signalling.


loudmouth/base64.c
/*
 * base64.c - Base 64 encoding/decoding implementation
 * Copyright (C) 2006 Collabora Ltd.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <ctype.h>
#include <string.h>

#include <glib.h>

#include "base64.h"

/*
|AAAA AABB|BBBB CCCC|CCDD DDDD|

0xFC = 1111 1100
0x03 = 0000 0011
0xF0 = 1111 0000
0x0F = 0000 1111
0xC0 = 1100 0000
0x3F = 0011 1111

3 input bytes = 4 output bytes;
2 input bytes = 2 output bytes;
1 input byte  = 1 output byte.
*/

static const gchar *encoding =
  "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

static const guint decoding[256] =
{
  /* ... */
   0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
   0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
   0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
   0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
   0, 0, 0,
  /* + */
  62,
  /* ... */
   0, 0, 0,
  /* / , 0-9 */
  63,
  52, 53, 54, 55, 56, 57, 58, 59, 60, 61,
  /* ... */
   0, 0, 0, 0, 0, 0, 0,
  /* A */
   0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12,
  13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
  /* ... */
   0, 0, 0, 0, 0, 0,
  /* a */
  26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
  39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51
};

#define GET_6_BITS_0(s) (((s)[0] & 0xFC) >> 2)
#define GET_6_BITS_1(s) (((s)[0] & 0x03) << 4) | \
                        (((s)[1] & 0xF0) >> 4)
#define GET_6_BITS_2(s) (((s)[1] & 0x0F) << 2) | \
                        (((s)[2] & 0xC0) >> 6)
#define GET_6_BITS_3(s) (((s)[2] & 0x3F) << 0)

#define GET_BYTE_0(s) (((decoding[(guchar)(s)[0]] & 0x3F) << 2) | \
                       ((decoding[(guchar)(s)[1]] & 0x30) >> 4))
#define GET_BYTE_1(s) (((decoding[(guchar)(s)[1]] & 0x0F) << 4) | \
                       ((decoding[(guchar)(s)[2]] & 0x3C) >> 2))
#define GET_BYTE_2(s) (((decoding[(guchar)(s)[2]] & 0x03) << 6) | \
                       ((decoding[(guchar)(s)[3]] & 0xFF) << 0))

gchar *_lm_base64_encode (const gchar *txt, gsize n)
{
  guint i;
  guint len;
  GString *tmp;
  GString *str = g_string_new_len (txt, n);

  len = str->len;
  /* TODO: calculate requisite output string length and allocate that big a
   * GString */
  tmp = g_string_new ("");

  for (i = 0; i < len; i += 3)
    {
      guint c1, c2, c3, c4;

      if (i > 0 && (i * 4) % 76 == 0)
          g_string_append_c (tmp, '\n');

      switch (i + 3 - len)
        {
        case 1:
          c1 = encoding[GET_6_BITS_0 (str->str + i)];
          c2 = encoding[GET_6_BITS_1 (str->str + i)];
          c3 = encoding[GET_6_BITS_2 (str->str + i)];
          c4 = '=';
          break;
        case 2:
          c1 = encoding[GET_6_BITS_0 (str->str + i)];
          c2 = encoding[GET_6_BITS_1 (str->str + i)];
          c3 = '=';
          c4 = '=';
          break;
        default:
          c1 = encoding[GET_6_BITS_0 (str->str + i)];
          c2 = encoding[GET_6_BITS_1 (str->str + i)];
          c3 = encoding[GET_6_BITS_2 (str->str + i)];
          c4 = encoding[GET_6_BITS_3 (str->str + i)];
        }

      g_string_append_printf (tmp, "%c%c%c%c", c1, c2, c3, c4);
    }

  return g_string_free (tmp, FALSE);
}

gchar *_lm_base64_decode (const gchar *str, gsize *len)
{
  guint i;
  GString *tmp;
  char group[4];
  guint filled = 0;

  *len = 0;

  for (i = 0; str[i]; i++)
    {
      if (str[i] != 'A' &&
          str[i] != '=' &&
          !isspace(str[i]) &&
          decoding[(guchar) str[i]] == 0)
        {
          g_debug ("bad character %x at byte %u", (guchar)str[i], i);
          return NULL;
        }
    }

  tmp = g_string_new ("");

  for (i = 0; str[i]; i++)
    {
      if (isspace(str[i]))
        continue;

      group[filled++] = str[i];

      if (filled == 4)
        {
          if (group[3] == '=')
            {
              if (group[2] == '=')
                {
                  g_string_append_c (tmp, GET_BYTE_0(group));
                }
              else
                {
                  g_string_append_c (tmp, GET_BYTE_0(group));
                  g_string_append_c (tmp, GET_BYTE_1(group));
                }
             }
           else
            {
              g_string_append_c (tmp, GET_BYTE_0(group));
              g_string_append_c (tmp, GET_BYTE_1(group));
              g_string_append_c (tmp, GET_BYTE_2(group));
            }
          filled = 0;
        }
    }

  if (filled)
    {
      g_debug ("insufficient padding at end of base64 string:\n%s", str);
      g_string_free (tmp, TRUE);
      return NULL;
    }

  *len = tmp->len;
  return g_string_free (tmp, FALSE);
}


